File: old experiment/javaSwapChain/src/SwapChain/GA.java
package SwapChain;

import java.util.Random;

public class GA {
    private Gene[] Population;//种群
    private int  PopulationSize;//种群大小
    private double Pm;//变异概率
    private double Pc;//交叉概率
    private double Ps;//选择概率
    private Providers P[];
    private Customer O[];
    private double D;
    private final double BigNum=21400000;
    private int IterMax;
    
    public GA(Providers p[],Customer o[],double d)
    //种群大小,选择概率，交叉概率，变异概率，P集合,O集合
    {
    	PopulationSize=100;
    	Pm=0.00001;
    	Pc=0.01;
    	Ps=0.01;
    	IterMax=50;
    	
    	Population = new Gene[PopulationSize];
    	P=p;
    	O=o;
    	D=d;
    	Gene.GeneLength=P.length;
    	
    	//这里只是声明了种群大小的引用,真正使用时，还要对每个对象进行初始化
    }
    
    public void Solver()
    {
    	initialize();
    	/*for (int i=0;i<PopulationSize;i++)
    	{
            String s;
            s=Double.toString(Population[i].fitness);
            
            System.out.println(s);
    	}*/
    	
    	int iter=0;
    	double theBestFitness=-2*BigNum;
    	int theBestSerial[]=new int[P.length];
    	while(iter < IterMax)
    	{
    		
    		replicate();
    		
    		cross();
    		
    		mutation();
    		
    		//统一重新计算适应度值
    		for (int i=0;i<PopulationSize;i++)
        		Population[i].fitness=calc_fitness(Population[i]);
    		
    		iter++;
    		
    		for (int i=0;i<PopulationSize;i++)
        	{
                String s;
                s=Double.toString(Population[i].fitness);
                if (Population[i].fitness>=theBestFitness)
                {
                	theBestFitness=Population[i].fitness;
                	for (int j=0;j<P.length;j++)
    					theBestSerial[j]=Population[i].GeneSerial[j];
                }
                
                double SigmaCost=0;
				for (int j=0;j<P.length;j++)
					SigmaCost+=P[j].cost[Population[i].GeneSerial[j]];
                s+=" "+ SigmaCost;
                System.out.println(s);
        	}
        	
        	System.out.println();
    	}
    	
    	System.out.println();
    	
//    	qsort(0,PopulationSize-1,Population);
//    	//只输出最后一个作为GA的最优解
//    	for (int i=PopulationSize-1;i<PopulationSize;i++)
//    	{
//            String s;
//            s=Double.toString(Population[i].fitness);
//            double SigmaCost=0;
//			for (int j=0;j<P.length;j++)
//				SigmaCost+=P[j].cost[Population[i].GeneSerial[j]];
//            s+=" "+ SigmaCost+" ";
//            for (int j=0;j<P.length;j++)
//			{
//			   s+=Integer.toString(Population[i].GeneSerial[j])+ " ";
//			}
//            System.out.println(s);
//    	}
    	String s;
      s=Double.toString(theBestFitness);
      double SigmaCost=0;
		for (int j=0;j<P.length;j++)
			SigmaCost+=P[j].cost[theBestSerial[j]];
      s+=" "+ SigmaCost+" ";
      for (int j=0;j<P.length;j++)
		{
		   s+=Integer.toString(theBestSerial[j])+ " ";
		}
      System.out.println(s);
    	System.out.println();
    }
    
    private void qsort(int st,int ed,Gene[] a)
    {
    	if (st>=ed)
    		return;
    	int i=st,j=ed;
    	Gene tmp=a[i];
    	while (i<j)
    	{
    		while (i<j && a[j].fitness>=tmp.fitness) j--;
    		if (i<j)
    			a[i]=a[j];
    		while (i<j && a[i].fitness<tmp.fitness) i++;
    		if (i<j)
    			a[j]=a[i];
    	}
    	a[i]=tmp;
    	qsort(st,i-1,a);
    	qsort(i+1,ed,a);
    	return;
    }
    
    private void initialize()
    //随机初始化种群
    {
    	Random rd=new Random();
    	for (int i=0;i<PopulationSize;i++)
    	{
    		//某一条染色体
    		Population[i]=new Gene(P.length);
    		for (int j=0;j<Gene.GeneLength;j++)
    			Population[i].GeneSerial[j]=rd.nextInt(P[j].CapacityCtr);
    		Population[i].fitness=calc_fitness(Population[i]);
    	}
    }
    
    private void replicate()
    {
    	Gene[] NextPopulation = new Gene[PopulationSize];//声明下一代种群
        for (int i = 0; i < PopulationSize; i++)//对种群中每个个体进行声明
            NextPopulation[i] = new Gene(Gene.GeneLength);

        double[] pi = new double[PopulationSize];
        double FitnessSum = 0;
        //复制
        FitnessSum = 0;
        qsort(0, PopulationSize - 1,Population);
        //采取“保留最佳精英策略”，将最优个体直接替代最差个体;
        //但最差个体仍有机会参与复制
        NextPopulation[0] = Population[PopulationSize - 1];
        //构造轮盘赌
        for (int i = 0; i < PopulationSize; i++)
            FitnessSum += Population[i].fitness;

        pi[0] = Population[0].fitness / FitnessSum;
        for (int i = 1; i < PopulationSize; i++)
            pi[i] = Population[i].fitness / FitnessSum + pi[i - 1];

        Random rd = new Random();
        for (int i = 1; i < PopulationSize; i++)
        {
            double tmp = rd.nextDouble();
            int copy = 0;
            for (int j = 0; j < PopulationSize; j++)
                if (tmp <= pi[j])
                { copy = j; break; }
            NextPopulation[i] = Population[copy];
        }

        for (int i = 0; i < PopulationSize; i++)
            Population[i] = NextPopulation[i];
    }
    
    private void cross()
    {
    	int[] hash = new int[PopulationSize];
        Random rd = new Random();
        //交叉
        for (int i = 0; i < PopulationSize; i++)
            hash[i] = 0;
        hash[0] = 1;
        //最佳个体不参与交叉操作
        for (int i = 1; i < PopulationSize / 2; i++)
        {
            hash[i] = 1;
            int j = 0;//hash[0] = 1;
            while (hash[j] == 1)
            {
                j = PopulationSize/2 + rd.nextInt(PopulationSize/2);
                //找到一个hash[j]==0的，就跳出
            }
            hash[j] = 1;
            //Xi,Xj进行交叉,交换特定的一段染色体
            
            /*//单点交叉
            int cross_point = rd.nextInt(Gene.GeneLength);
            if (i != 0 && j != 0)
            {
               int tmp;
               for (int k=cross_point;k<Gene.GeneLength;k++)
               {
            	   tmp=Population[i].GeneSerial[k];
            	   Population[i].GeneSerial[k]=Population[j].GeneSerial[k];
            	   Population[j].GeneSerial[k]=tmp;
               }
            }*/
            
            //两点交叉
            int cross_point1 = rd.nextInt(Gene.GeneLength);
            int cross_point2 = rd.nextInt(Gene.GeneLength);
            if (i != 0 && j != 0)
            {
               int tmp;
               for (int k=cross_point1;k<=cross_point2;k++)
               {
            	   tmp=Population[i].GeneSerial[k];
            	   Population[i].GeneSerial[k]=Population[j].GeneSerial[k];
            	   Population[j].GeneSerial[k]=tmp;
               }
            }
        }
    }
    
    private void mutation()
    {
    	Random rd = new Random();
        //变异
        for (int k = 0; k < PopulationSize * Gene.GeneLength * Pm; k++)
        {
            //int i = rd.Next(0, PopulationSize);
            int i = rd.nextInt(PopulationSize);//随机抽取个体Xi
            //int ik = rd.Next(0, GENE.GeneLength);//
            int ik = rd.nextInt(Gene.GeneLength);//随机选取需要变异的基因位
            int vk = rd.nextInt(P[ik].CapacityCtr);//产生变异值
            
            Population[i].GeneSerial[ik] = vk;
        }
    
    }
    
    private double calc_fitness(Gene one)
    {
    	double alpha=5.0,beta=1.0;
    	//以上为可调的两个因子
    	double fitness=0;
    	Provider tmpP[]=new Provider[P.length];
    	double SigmaCapacity=0,SigmaDemand=0;
		for (int i=0;i<P.length;i++)
		{
			tmpP[i]=new Provider();
			tmpP[i].x=P[i].x;
			tmpP[i].y=P[i].y;
			tmpP[i].Capacity=P[i].capacity[one.GeneSerial[i]];
			SigmaCapacity+=tmpP[i].Capacity;
			/*String s = Double.toString(tmpP[i].x) + " " +
					   Double.toString(tmpP[i].y) + " " +"1 "+
					   Double.toString(tmpP[i].Capacity);
			System.out.println(s);*/
		}
		
		Customer tmpO[]=new Customer[O.length];
		for (int i=0;i<O.length;i++)
		{
			tmpO[i]=new Customer();
			tmpO[i].x=O[i].x;
			tmpO[i].y=O[i].y;
			tmpO[i].Demand=O[i].Demand;
			SigmaDemand+=tmpO[i].Demand;
		}
		
		if (SigmaCapacity<SigmaDemand)
			fitness=-2*BigNum;
		else
		{
			SwapChainSolver scs = new SwapChainSolver(tmpP, tmpO);
			double mmd = scs.Solver();
			
			//调试输出
			/*System.out.printlUn(mmd);*/
			/*String s= new String();
			for (int i=0;i<P.length;i++)
			{
			   s+=Integer.toString(one.GeneSerial[i])+ " ";
			}
			System.out.println(s);*/
			
			if (mmd>D)
				fitness=-BigNum;
			else
			{
				double SigmaCost=0;
				for (int i=0;i<P.length;i++)
					SigmaCost+=P[i].cost[one.GeneSerial[i]];
				fitness=-alpha*SigmaCost+beta*(D-mmd);
			}
		}
		 
        return fitness;
    }
    
}

File: new experiment/code/GA.py
import math
import random
from random import randint
import copy
from sklearn.metrics import mean_squared_error
from Surrogate import Surrogate
from SwapChainSolver import SwapChainSolver
from PO import PO
from Customer import Customer
from Provider import Provider
from ProviderPlus import ProviderPlus
import numpy as np
from Chromosome import Chromosome
from FileProcess import *

class GA:
    def __init__(self, maxIter, maxBlock, populationSize, probMutate, probCross, probSelect, D, po, alpha, beta, surrogateFlag):
        self.m_MaxIter = maxIter
        self.m_MaxBlock = maxBlock
        self.m_PopulationSize = populationSize
        self.m_Population = []
        self.m_SurrogateFlag = surrogateFlag

        self.m_ProbMutate = probMutate
        self.m_ProbCross = probCross
        self.m_ProbSelect = probSelect
        self.m_PO = po
        self.m_D = D
        self.m_Alpha = alpha
        self.m_Beta = beta
        self.m_Block = 0

        for _ in range(self.m_PopulationSize):
            self.m_Population.append(self.generateRandomChromosome())
        tmp = sorted(self.m_Population, key=lambda x:x.fitness, reverse=True)
        self.m_BestSolution = tmp[0]
        self.m_BestFitness = self.m_BestSolution.fitness

    def select(self):
        nextPopulation = []
        pi = []
        fitnessSum = 0
        self.m_Population = sorted(self.m_Population, key=lambda x:x.fitness)
        nextPopulation.append(copy.deepcopy(self.m_Population[-1]))

        for ind in self.m_Population:
            fitnessSum = fitnessSum + ind.fitness
        pi.append(self.m_Population[0].fitness / fitnessSum)
        for ri in range(1, len(self.m_Population)):
            pi.append(self.m_Population[ri].fitness / fitnessSum + pi[ri - 1])
        copyNum = len(self.m_Population) - 1
        for ri in range(1, len(self.m_Population)):
            randnum = random.random()
            for j in range(len(pi)):
                if randnum <= pi[j]:
                    copyNum = j
                    break
            nextPopulation.append(copy.deepcopy(self.m_Population[copyNum]))
        self.m_Population = nextPopulation

    def crossover(self):
        # chromosomes cross
        hash = []
        for ci in range(len(self.m_Population)):
            hash.append(0)
        hash[0] = 1
        for ci in range(1, len(self.m_Population) / 2):
            hash[ci] = 1
            j = 0
            while hash[j] == 1:
                j = len(self.m_Population) / 2 + randint(0, len(self.m_Population) / 2 - 1)
            hash[j] = 1
            if random.random() > self.m_ProbCross:
                # cross gene between pointA and pointB
                pointA = randint(0, len(self.m_Population[0].geneSerial) - 1)
                pointB = randint(0, len(self.m_Population[0].geneSerial) - 1)
                if pointA >= pointB:
                    tmp = pointA
                    pointA = pointB
                    pointB = tmp
                if ci != 0 and j != 0:
                    for k in range(pointA, pointB + 1):
                        tmp = self.m_Population[ci].geneSerial[k]
                        self.m_Population[ci].geneSerial[k] = self.m_Population[j].geneSerial[k]
                        self.m_Population[j].geneSerial[k] = tmp

    def mutate(self):
        # chromosomes mutation
        for k in range(0, int(len(self.m_Population) * len(self.m_Population[0].geneSerial) * self.m_ProbMutate) + 1):
            mi = randint(0, len(self.m_Population) - 1)
            ik = randint(0, len(self.m_Population[0].geneSerial) - 1)
            vk = randint(0, self.m_PO.PROVIDERS[ik].cnt - 1)
            self.m_Population[mi].geneSerial[ik] = vk

    def calcPopulationFitness(self):
        for chromosome in self.m_Population:
            if self.m_SurrogateFlag:
                chromosome.fitness, chromosome.mmd, chromosome.sigmaCapacity, chromosome.sigmaCost, chromosome.sigmaDemand = self.calcFitnessWithSurrogate(
                    chromosome.geneSerial, self.m_PO, self.m_D)
            else:
                chromosome.fitness, chromosome.mmd, chromosome.sigmaCapacity, chromosome.sigmaCost, chromosome.sigmaDemand = self.calcFitness(
                    chromosome.geneSerial, self.m_PO, self.m_D)


    def evaluate(self):
        iter = 0
        while iter < self.m_MaxIter: #and self.m_Block < self.m_MaxBlock:
            #print "the " + str(iter) + " th iteration"
            self.select()
            self.crossover()
            self.mutate()
            self.calcPopulationFitness()
            sortedPopulation = sorted(self.m_Population, key=lambda x: x.fitness, reverse=True)
            if sortedPopulation[0].fitness > self.m_BestFitness:
                self.m_BestFitness = sortedPopulation[0].fitness
                self.m_BestSolution = copy.deepcopy(sortedPopulation[0])
                self.m_Block = 0
            elif math.fabs(sortedPopulation[0].fitness - self.m_BestFitness) <= 0.00001:
                self.m_Block += 1

            #print "the best individual serial, fitness, mmd, sigmaCost, sigmaCapacity, sigmaDemand ",\
            #    sortedPopulation[0].geneSerial, sortedPopulation[0].fitness,sortedPopulation[0].mmd, sortedPopulation[0].sigmaCost, sortedPopulation[0].sigmaCapacity, sortedPopulation[0].sigmaDemand
            print sortedPopulation[0].sigmaCost

            iter += 1


    def generateRandomChromosome(self):
        chromosome = Chromosome()
        for i in range(len(self.m_PO.PROVIDERS)):
            chromosome.geneSerial.append(randint(0, self.m_PO.PROVIDERS[i].cnt - 1))
        if self.m_SurrogateFlag:
            chromosome.fitness, chromosome.mmd, chromosome.sigmaCapacity, chromosome.sigmaCost, chromosome.sigmaDemand = self.calcFitnessWithSurrogate(
                chromosome.geneSerial, self.m_PO, self.m_D)
        else:
            chromosome.fitness, chromosome.mmd, chromosome.sigmaCapacity, chromosome.sigmaCost, chromosome.sigmaDemand = self.calcFitness(
                chromosome.geneSerial, self.m_PO, self.m_D)
        return chromosome

    def calcFitness(self, geneSerial, data, D):
        """
            usage ChromosomeNumber,geneSerial,data,D
            return fitness for this1  Chromosome
        """
        alpha = self.m_Alpha
        beta = self.m_Beta
        # alpha and beta are weight factor
        customers = []
        fitness = 0
        for item in data.CUSTOMERS:
            tmp = Customer()
            tmp.x = copy.deepcopy(item.x)
            tmp.y = copy.deepcopy(item.y)
            tmp.demand = copy.deepcopy(item.demand)
            customers.append(tmp)
        providers = []
        sigmaCost = 0
        sigmaCapacity = 0
        sigmaDemand = 0
        mmd = self.m_D * 1000.0
        for i in range(0, len(geneSerial)):
            tmpProvider = Provider()
            tmpProvider.x = copy.deepcopy(data.PROVIDERS[i].x)
            tmpProvider.y = copy.deepcopy(data.PROVIDERS[i].y)
            tmpProvider.capacity = copy.deepcopy(data.PROVIDERS[i].capacity[geneSerial[i]])
            tmpProvider.cost = copy.deepcopy(data.PROVIDERS[i].cost[geneSerial[i]])
            sigmaCost = sigmaCost + tmpProvider.cost
            sigmaCapacity = sigmaCapacity + tmpProvider.capacity
            providers.append(tmpProvider)
        for item in customers:
            sigmaDemand = sigmaDemand + item.demand

        if sigmaCapacity >= sigmaDemand:
            swapchainsolver = SwapChainSolver(providers, customers)
            mmd = swapchainsolver.Solver()
            if mmd > D:
                fitness = -4.0
            else:
                if sigmaCost != 0:
                    fitness = float(4.0 / sigmaCost)
                else:
                    fitness = 8.0
        else:
            fitness = -8.0
        # print("fitness,mmd,sigmaCapacity,sigmaCost,sigmaDemand:",fitness,mmd,sigmaCapacity,sigmaCost,sigmaDemand)
        return math.exp(fitness), mmd, sigmaCapacity, sigmaCost, sigmaDemand

    def calcFitnessWithSurrogate(self, geneSerial, data, D):
        """
            usage ChromosomeNumber,geneSerial,data,D
            return fitness for this1  Chromosome
        """
        alpha = self.m_Alpha
        beta = self.m_Beta
        # alpha and beta are weight factor
        customers = []
        fitness = 0
        for item in data.CUSTOMERS:
            tmp = Customer()
            tmp.x = copy.deepcopy(item.x)
            tmp.y = copy.deepcopy(item.y)
            tmp.demand = copy.deepcopy(item.demand)
            customers.append(tmp)
        providers = []
        sigmaCost = 0
        sigmaCapacity = 0
        sigmaDemand = 0
        mmd = self.m_D * 1000.0
        for i in range(0, len(geneSerial)):
            tmpProvider = Provider()
            tmpProvider.x = copy.deepcopy(data.PROVIDERS[i].x)
            tmpProvider.y = copy.deepcopy(data.PROVIDERS[i].y)
            tmpProvider.capacity = copy.deepcopy(data.PROVIDERS[i].capacity[geneSerial[i]])
            tmpProvider.cost = copy.deepcopy(data.PROVIDERS[i].cost[geneSerial[i]])
            sigmaCost = sigmaCost + tmpProvider.cost
            sigmaCapacity = sigmaCapacity + tmpProvider.capacity
            providers.append(tmpProvider)
        for item in customers:
            sigmaDemand = sigmaDemand + item.demand

        if sigmaCapacity >= sigmaDemand:
            x = np.array(geneSerial).reshape(1, -1)
            mmd = self.m_Surrogate.predict(x)[0]
            if mmd > D:
                fitness = -1000
            elif mmd > 0:
                if sigmaCost != 0:
                    fitness = float(4.0 / sigmaCost)
                else:
                    fitness = 8.0
            else:
                fitness = -6.0
        else:
            fitness = -8.0
        # print"fitness,mmd,sigmaCapacity,sigmaCost,sigmaDemand:",fitness,mmd,sigmaCapacity,sigmaCost,sigmaDemand
        return math.exp(fitness), mmd, sigmaCapacity, sigmaCost, sigmaDemand

if __name__ == "__main__":
    # po is data contains informantion about PROVIDERS and CUSTOMERS
    po = PO()
    # read providers and customers data from text
    po.PROVIDERS, po.CUSTOMERS = LoadDataFromText(r"data6.txt")

    maxIter = 100
    maxBlock = 5
    populationSize = 300
    probMutate = 0.00001
    probCross = 0.7
    probSelect = 0.1
    D = 40.0
    alpha = 10000000.00
    beta = 0.01
    surrogateFlag = False

    ga = GA(maxIter,maxBlock,populationSize,probMutate,probCross,probSelect,D,po,alpha,beta,surrogateFlag)
    ga.evaluate()
    print "the best solution serial, fitness, mmd, sigmaCost, sigmaCapacity, sigmaDemand ", \
        ga.m_BestSolution.geneSerial, ga.m_BestSolution.fitness, ga.m_BestSolution.mmd, ga.m_BestSolution.sigmaCost, ga.m_BestSolution.sigmaCapacity, ga.m_BestSolution.sigmaDemand


